beacon-chain/verification/data_column.go

```diff
@@ -0,0 +1,504 @@
+	return nil
+}
+
+func (dv *RODataColumnsVerifier) SidecarProposerExpected(ctx context.Context) (err error) {
+	if ok, err := dv.results.cached(RequireSidecarProposerExpected); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireSidecarProposerExpected, &err)
+
+	for _, dataColumn := range dv.dataColumns {
+		// Extract the slot of the data column.
+		dataColumnSlot := dataColumn.Slot()
+
+		// Compute the epoch of the data column slot.
+		dataColumnEpoch := slots.ToEpoch(dataColumnSlot)
+		if dataColumnEpoch > 0 {
+			dataColumnEpoch = dataColumnEpoch - 1
+		}
+
+		// Extract the root of the parent block corresponding to the data column.
+		parentRoot := dataColumn.ParentRoot()
+
+		// Compute the target root for the epoch.
+		targetRoot, err := dv.fc.TargetRootForEpoch(parentRoot, dataColumnEpoch)
+		if err != nil {
+			return columnErrBuilder(errors.Wrap(err, "target root for epoch"))
+		}
+
+		// Create a checkpoint for the target root.
+		checkpoint := &forkchoicetypes.Checkpoint{Root: targetRoot, Epoch: dataColumnEpoch}
+
+		// Try to extract the proposer index from the data column in the cache.
+		idx, cached := dv.pc.Proposer(checkpoint, dataColumnSlot)
+
+		if !cached {
+			// Retrieve the parent state.
+			parentState, err := dv.parentState(ctx, dataColumn)
+			if err != nil {
+				return columnErrBuilder(errors.Wrap(err, "parent state"))
+			}
+
+			idx, err = dv.pc.ComputeProposer(ctx, parentRoot, dataColumnSlot, parentState)
+			if err != nil {
+				return columnErrBuilder(errors.Wrap(err, "compute proposer"))
+			}
+		}
+
+		if idx != dataColumn.ProposerIndex() {
+			return columnErrBuilder(ErrSidecarUnexpectedProposer)
+		}
+	}
+
+	return nil
+}
+
+// parentState retrieves the parent state of the data column from the cache if possible, else retrieves it from the state by rooter.
+func (dv *RODataColumnsVerifier) parentState(ctx context.Context, dataColumn blocks.RODataColumn) (state.BeaconState, error) {
+	parentRoot := dataColumn.ParentRoot()
+
+	// If the parent root is already in the cache, return it.
+	if st, ok := dv.stateByRoot[parentRoot]; ok {
+		return st, nil
+	}
+
+	// Retrieve the parent state from the state by rooter.
+	st, err := dv.sr.StateByRoot(ctx, parentRoot)
+	if err != nil {
+		return nil, errors.Wrap(err, "state by root")
+	}
+
+	// Store the parent state in the cache.
+	dv.stateByRoot[parentRoot] = st
+
+	return st, nil
+}
+
+func columnToSignatureData(d blocks.RODataColumn) SignatureData {
+	return SignatureData{
+		Root:      d.BlockRoot(),
+		Parent:    d.ParentRoot(),
+		Signature: bytesutil.ToBytes96(d.SignedBlockHeader.Signature),
+		Proposer:  d.ProposerIndex(),
+		Slot:      d.Slot(),
+	}
+}
+
+func columnErrBuilder(baseErr error) error {
+	return errors.Wrap(baseErr, errColumnsInvalid.Error())
+}
+
+func inclusionProofKey(c blocks.RODataColumn) ([32]byte, error) {
+	var buf bytes.Buffer
+
+	r, err := c.SignedBlockHeader.HashTreeRoot()
```

Comment:
What if we computed this when we create a RODataColumn and avoid ever recomputing it?

Reply:
Oh woow, I just missed the whole point of using a 'RODataColumn' here ;)

Fixed in 2857c2178147d94f0a8d2e18e800dba750f64c59.